napari/layers/shapes/_shapes_mouse_bindings.py

```diff
@@ -257,6 +257,8 @@ def _add_line_rectangle_ellipse(
         String indicating the type of shape to be added.
     """
     # on press
+    # reset layer._aspect_ratio for a new shape
+    layer._aspect_ratio = 1
```

Comment:
Where does the aspect ratio get set/changed? Why does the aspect ratio disable the shift behaviour? (Maybe add a comment around the relevant if-statement below.)

Reply:
_aspect_ratio is set in the __init__
and then when Shift is pressed:
https://github.com/napari/napari/blob/b87e9c7e430b5ff98225a86b92233ca913a95ff0/napari/layers/shapes/_shapes_key_bindings.py#L19-L40

Another option would be to set is back to 1 after the yield, so on shift release?
This makes sense, because _fixed_aspect is unset too...

Reply:
Huh, that didn't work. Pressing Shift sets it, but adding layer._aspect_ratio = 1 at the end doesn't reset is back to 1. Oh, wait it does, but then when you press Shift again to draw the next line it just resets the aspect ratio again.
This is all born out of the same code being used for manipulating selected objects -- drawing shapes makes the shape selected, so then it's all the same stuff.
I think reseting on draw is the correct way.

Reply:
Could do the reset just for add_line i guess?

Reply:
Would adding to the comment something like :
"so Shift doesn't pick up the _aspect_ratio of the previous shape" 
help?

Reply:
I think ultimately we should change the variable. It made sense when this was *only* being used for box resizing, but now we are also using it for rotation, it should probably have a more descriptive name... And be a context manager maybe? Dunno.

But the other part of the question was *why* the line snap mode should only work when aspect_ratio is 1?

But anyway, I'm happy for the PR to go in as-is and then to make an issue for rejiggering it.

Reply:
> Would adding to the comment something like :
> "so Shift doesn't pick up the _aspect_ratio of the previous shape"

It would, but potentially a longer comment would help even more. Something like,

"when resizing a shape while holding shift, we maintain the aspect ratio of the shape at the time shift was pressed. This resets the ratio so that resizing a *new* shape doesn't snap to the aspect ratio of the previous resizing attempt."

But I gotta say as I wrote that it really felt like the resetting should happen in the shift-calling code...

I also wonder if None is a better value here...

Reply:
Oh, to preserve existing behavior, where if you start drawing a line *then press shift* you lock the *current* aspect ratio of the line. This is really handy.
So basically, if you have aspect ratio 1, the layer _init_ you're at default. Press shift and you preserve that, but let lines rotate through 45 steps -- same as the transform of selected of shapes/layers.
If you have some other aspect ratio, you preserve that, so you can draw long lines of a specific angle easily.

Reply:
This works just fine if you have no other shapes selected when you press shift. 
I guess we could do some dettection of mode in the key binding instead of in the mouse binding code.